### ISSUE TITLE
Refactor Scripts/CfgParser into a hemeTools module

### ISSUE THREAD

**Issue by schmie** [schmie/hemeMove#47]
**Reported by rupert on 6 Oct 2011 15:30 UTC**
The !CfgParser is useful, but the code was hard to reuse. Here I have refactored the loading into memory of a config file into an event-driven parser (`hemeTools.parsers.config.simple.ConfigLoader`). A subclass (`hemeTools.parsers.config.freeing.FreeingConfigLoader`) allows processing to occur as soon as a block has its neighbours available and blocks are deleted when no longer needed.

A further subclass does the processing asynchronously and in parallel.

All that remains in !CfgParser.py is the code to check a single Block for validity.

Usage: ensure hemeTools is installed (`python setup.py install [--user]`) then run as before (`python CfgParser.py /path/to/config.dat`)

Known issues: killing the program with ctrl-c doesn't always give a clean exit because of the child processes. Workaround: kill -9 the child processes

**Comment by schmie** [comment]
**Comment by hywel on 7 Oct 2011 10:30 UTC**
Apparently, if you make child processes daemonic they will be killed when the parent is. Can you try this out? If that doesn't work / isn't viable, you could catch SIGTERM and iterate over all the child processes and kill them from within the Python. http://stackoverflow.com/questions/2542610/python-daemon-doesnt-kill-its-kids

Have you tested on only valid files? It would be good to test on at least one invalid file to make sure the validity checks are actually being run.

Have you also checked that the memory usage stays constant? I mainly ask because it looks like errors are only printed once at the end. If the process dies before completion / the user is impatient, we'll never know if the file was valid. Could you instead print out the errors every BLOCK_REPORT_FREQ iterations, then clear the error queue? 

I'm also not clear on how you can identify the block / site that's at fault at the moment, without using pdb. No doubt the most important thing is knowing if it's valid / invalid but you can imagine how much more helpful it'd be to know exactly why it's invalid, to help with debugging etc. (See 'Site doesn't appear to have any fitting type', 'Site is non-normal fluid but has all valid fluid neighbours.'). Possibly this is stuff you're expecting me to fix when I rewrite some of the constraint-checking code.

The comment beginning '# Now we iterate over every site on every block, and check the following constraints.' needs to be moved to the start of the constraint-checking.

Your __init__.py has only comments, no actual script. Is that intentional?

Why not "PRESSURE_EDGE_MASK = 1 << 31"?

Why does the !FreedBlock class remain but is commented out? Ditto the !BlocksThatNeedBlock array.

Is it intentional that you're counting each block as a neighbour of itself? If so, this deserves a comment in the code.

Why can't you call !OnBlockProcessed after calling !OnBlockNeighboursAvailable rather than relying on the derived class to call it at the end of its implementation? If there's a good reason, then you should explain this in a comment.

Otherwise, nice.

**Comment by schmie** [comment]
**Comment by hywel on 7 Oct 2011 10:32 UTC**
Also, I can't work out which bit I'm meant to be rewriting; it might be easier for you to do. If it's the constraint-checking then I think the output strings should tell you the intention of each test.

**Comment by schmie** [comment]
**Comment by rupert on 7 Oct 2011 14:35 UTC**
Replying to [hywel](comment:1):
> Apparently, if you make child processes daemonic they will be killed when the parent is. Can you try this out? If that doesn't work / isn't viable, you could catch SIGTERM and iterate over all the child processes and kill them from within the Python. http://stackoverflow.com/questions/2542610/python-daemon-doesnt-kill-its-kids
This is the wrong approach here. I have found a better one: tell the children to ignore SIGINT and rely on the main process to terminate its children when the Pool is destructed.

> Have you tested on only valid files? It would be good to test on at least one invalid file to make sure the validity checks are actually being run.
Will do before pushing, although from debugging I am certain the checks are run, even if i'm not certain they are run correctly...

> Have you also checked that the memory usage stays constant? I mainly ask because it looks like errors are only printed once at the end. If the process dies before completion / the user is impatient, we'll never know if the file was valid. Could you instead print out the errors every BLOCK_REPORT_FREQ iterations, then clear the error queue? 

I have now, and fixed a problem to do with circular references meaning that Blocks are not free ASAP (they were only freed periodically by the garbage collector). You reasoning about error printing is incorrect: they are printed once the block has been fully checked and the results have to to the top of the Pool's results queue.

> I'm also not clear on how you can identify the block / site that's at fault at the moment, without using pdb. No doubt the most important thing is knowing if it's valid / invalid but you can imagine how much more helpful it'd be to know exactly why it's invalid, to help with debugging etc. (See 'Site doesn't appear to have any fitting type', 'Site is non-normal fluid but has all valid fluid neighbours.'). Possibly this is stuff you're expecting me to fix when I rewrite some of the constraint-checking code.
This info was not given in the original code. It can certainly be added, but not just yet
 
> The comment beginning '# Now we iterate over every site on every block, and check the following constraints.' needs to be moved to the start of the constraint-checking.
Fixed. 

> Your __init__.py has only comments, no actual script. Is that intentional?
Yes - it's a common idiom 
 
> Why not "PRESSURE_EDGE_MASK = 1 << 31"?
FIxed
 
> Why does the !FreedBlock class remain but is commented out? Ditto the !BlocksThatNeedBlock array.
Gone
> Is it intentional that you're counting each block as a neighbour of itself? If so, this deserves a comment in the code.
Yes, and done.

> Why can't you call !OnBlockProcessed after calling !OnBlockNeighboursAvailable rather than relying on the derived class to call it at the end of its implementation? If there's a good reason, then you should explain this in a comment.
To enable asynchronous processing of Blocks. Explained in class's docstring.

> Otherwise, nice.

**Comment by schmie** [comment]
**Comment by rupert on 7 Oct 2011 15:05 UTC**
Running on a deliberately borked file (a simple fluid site surrounded by simple fluid, when set to solid) gives:
```
Checked block 0 of 306
ERROR: Fluid site at [ 4  5 55] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 4  7 55] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 5  6 55] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 6  5 55] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 6  7 55] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 4  5 57] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 4  6 56] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 4  7 57] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 5  5 56] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 5  6 57] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 5  7 56] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 6  5 57] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 6  6 56] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site at [ 6  7 57] has type, edge of 1, False and has some neighbours that aren't valid or are solids, e.g. site at [ 5  6 56]with type 0
ERROR: Fluid site counts not equal.
Checked block 100 of 306
Checked block 300 of 306
Checked block 200 of 306
```
Resolved with changeset c510de109103
